extensions/amp-form/0.1/amp-form.js

```diff
@@ -177,8 +185,69 @@ export class AmpForm {
 
 
 /**
+ * Reports validity for the first invalid input - if any.
+ * @param {!HTMLFormElement} form
+ * @private
+ */
+function reportFormValidity(form) {
+  const inputs = toArray(form.querySelectorAll('input,select,textarea'));
```

Comment:
very anecdotal but :scope has always been faster for me because querySelectorAll is not element rooted in the way it works. (it does not work the way you think it does)

see http://ejohn.org/blog/thoughts-on-queryselectorall/ for explanation.